### PR TITLE
New subscription system for reference cursors

### PR DESCRIPTION
**Pull request by tomasd**
Reference cursors are now notified about the change also in case when something on the path to the reference is changed.

### CODE REVIEW

**Comment by mikaelbr** [comment]
Thanks! I'll look more into this and test it out locally today, and merge if everything looks good. I should also probably do a performance test.

@andrewluetgers how about your test here?

**Comment by torgeir** [comment]
This ought to allow for someting like https://github.com/omniscientjs/immstruct/issues/19, no?

File: tests/structure_test.js [hunk 1]
```diff
@@ -1039,7 +1059,7 @@ describe('structure', function () {
         ref.observe(function () { i++; });
         ref2.observe(function () { i2++; });
         ref2.cursor().update(function () { return 'updated'; });
-        ref2.observe(function () { if(i == 1 && i2 == 2) done(); });
+        ref2.observe(function () { if(i == 1 && i2 == 1) done(); });
```
Comment (andrewluetgers) [inline tests/structure_test.js]: Why was this changed? Does it not invalidate the whole purpose of this test?

Comment (mikaelbr) [inline tests/structure_test.js]: Good catch! `unobserveAll()` should only remove listeners from the given reference, not other references created.

File: tests/structure_test.js [hunk 2]
```diff
@@ -849,6 +849,26 @@ describe('structure', function () {
...
+      });
+
+      it('should trigger change listener for depending observed references', function (done) {
+        var structure = new Structure({
+          data: {a: {b: {c: "value"}}}
+        });
+
+        var ref = structure.reference(['a', 'b']);
+        ref.observe(function () { done(); });
+        structure.cursor().update(function (old) { return old.setIn(['a', 'b', 'c'], 'updated'); });
```
Comment (Dashed) [inline tests/structure_test.js]: Is there a fundamental difference to updating via the root and via a sub-tree?

File: src/structure.js [hunk 3]
```diff
@@ -493,35 +512,12 @@ function hasIn(cursor, path) {
...
-    listeners = listeners.setIn(path, matched);
-  }
-
-  return {
-    matched: matched,
-    listeners: listeners
-  };
+function pathCombinations(path) {
+  var paths = (path||[]).reduce(function(acc,segment) {
+    acc.push(acc[acc.length-1].concat(segment));
```
Comment (Dashed) [inline src/structure.js]: I'm wondering about the performance impact of this.

File: tests/structure_test.js (hunk 1 above)
Comment (tomasd) [inline tests/structure_test.js]: Sorry, I've missed the last line of the test :-). As I've used Set as a container for listener methods, the order was not guaranteed and callback with done was called before i2++. Set was changed to OrderedSet

File: tests/structure_test.js (hunk 2 above)
Comment (tomasd) [inline tests/structure_test.js]: Basically no, I just wanted to be sure that root is handled the same way as subroot

File: src/structure.js (hunk 3 above)
Comment (tomasd) [inline src/structure.js]: I've just changed it into Immutable.List

**Comment by andrewluetgers** [comment]
@tomasd does this solve #37 ?

**Comment by tomasd** [comment]
@andrewluetgers Yes, it should be

**Comment by Dashed** [comment]
@andrewluetgers @tomasd No. It doesn't look like it.

To resolve https://github.com/omniscientjs/immstruct/issues/37, something like this would need to be in the test:
```js
        var ref2 = structure.reference(['a', 'b', 'c']);
        ref2.observe(function () { done(); });
```

**Comment by andrewluetgers** [comment]
Looks like invoke is dead code now

**Comment by andrewluetgers** [comment]
Sorry to say it but there is a significant performance regression. This test also does not deal with deeply nested structures so it is not well suited to really put this change to the test, I would assume perf would be even worse with deeply nested structures? At any rate I will need to revamp this test to do just that so we can know for sure.

with current master ~170 fps
http://plnkr.co/edit/upGq3HAqeDW6RHgo96IZ?p=preview


with latest changes ~5 fps
http://plnkr.co/edit/RdGpH5v6XOnYnvID6w9w?p=preview

**Comment by andrewluetgers** [comment]
One suggestion, If there is not fundamental need for using an immutable structure anywhere in your changes, best to opt for native mutable js alternatives.

**Comment by Dashed** [comment]
The performance regression isn't surprising since laziness/shortcut fusion isn't being utilized.

**Comment by andrewluetgers** [comment]
@Dashed are you implying laziness was being used before or could be? I'm not sure there is much of an opportunity for it here but I'm no expert on that. 

Took a look at how it scales as you change the size of the matrix. 10x10 @ 140 fps, 20x20 @ 88 fps, 30x30 @ 44 fps, 40x40 @ 30 fps

File: src/structure.js [hunk 4]
```diff
@@ -66,12 +66,13 @@ function Structure (options) {
       Infinity;
   }
 
+   this._referencelisteners = Immutable.Map();
+
   this._pathListeners = Immutable.Map();
```
Comment (mikaelbr) [inline src/structure.js]: Still not needed (was re-added in new PR)

File: src/structure.js [hunk 5]
```diff
@@ -66,12 +66,13 @@ function Structure (options) {
...
   this.on('swap', function (newData, oldData, keyPath) {
-    var listenerData = listMatchingOrCreateEmptyList(self._pathListeners, keyPath);
-    self._pathListeners = listenerData.listeners;
-
-    listenerData.matched.forEach(function (fn) {
+    pathCombinations(keyPath).reduce(function(fns, path) {
+      return fns.concat(self._referencelisteners.get(path, Immutable.Set()));
+    }, Immutable.Set()).forEach(function(fn) {
```
Comment (mikaelbr) [inline src/structure.js]: Here we potentially iterate twice, and that might be a problem if the `pathCombinations` is a large list and the result of the reduce is as well. Here for instance, you could do this in a single step (or laziness)

File: src/structure.js [hunk 6]
```diff
@@ -493,35 +518,11 @@ function hasIn(cursor, path) {
...
-  });
-}
-
-function listMatchingOrCreateEmptyListOnNamespace (listeners, path) {
-  path = (path || []).concat(LISTENER_SENTINEL);
-  var matched = listeners.getIn(path);
-
-  if (!matched) {
-    matched = [];
-    listeners = listeners.setIn(path, matched);
```
Comment (andrewluetgers) [inline src/structure.js]: I think this is pretty key, it represents the whole theory of the perf gains we accomplished with #55 that idea being having a tree structure to store callbacks and traversing that rather than everything to get the functions to call. Which it appears has been wiped out in these changes.

**Comment by mikaelbr** [comment]
I suspected as much. As this iterate much more, it might in cases where there are many listeners get slow. Where we need to improve the performance is at `on swap` and inside `pathCombinations`. Those operations are the ones who get called a lot.

File: src/structure.js (hunk 6 above)
Comment (andrewluetgers) [inline src/structure.js]: the getIn and setIn calls being titular here

**Comment by mikaelbr** [comment]
If we were to fix the performance issue and extend it to solve #37, it'd be a pretty good day! Will make immstruct a lot more powerful and versatile.

**Comment by andrewluetgers** [comment]
I agree, I think there is a different approach to this that can work, Im playing around with it. I'm thinking something like this would do it.
```javascript
  this.triggerObservers = function(newData, oldData, keyPath) {
   var listenerData = listMatchingOrCreateEmptyList(self._pathListeners, keyPath);
   self._pathListeners = listenerData.listeners;
    
   listenerData.matched.forEach(function (fn) {
     fn(keyPath, newData, oldData);
   });
  }

  this._pathListeners = Immutable.Map();
   
  this.on('swap', function (newData, oldData, keyPath) {
    self.triggerObservers(newData, oldData, keyPath);
    var parentKeyPath = // pop off last keyPath item
    var parent =  // get parent from parentKeyPath
    parent.triggerObservers();
  });
```

**Comment by tomasd** [comment]
@andrewluetgers Thank you for the performance report. I'll try to fix it soon.
Regarding #37 I've tried example from TLDR and it worked:
```
describe('xxx', function() {
     it('xxx', function(done) {
         var structure = Structure({
             data: {  }
         });

// this doesn't work as intended. very unexpected when user doesn't know how
// Cursor.from(...) works.
         structure
             .reference(['config', 'profile'])
             .observe(function(newObj, oldObj, hotKeyPath) {
                 // no output
                 done();
             });


// update some object
         structure.cursor().set('config', Immutable.fromJS({
             'profile': 'throwaway',
             'color': 'green'
         }));
     })  ;
})  ;
```

Or is there any other problem which I've missed? Thanks

**Comment by Dashed** [comment]
Ah. I've misread the code for the `subscribe` function. Yes, you're right, it resolves https://github.com/omniscientjs/immstruct/issues/37.

----

However, after looking at the code more closely, there are some things that doesn't seem right.

It seems you're populating `this._referencelisteners` in a lookup table manner:
```js
{ 'List []': [ [Function: cursorRefresher], [Function: listener] ],
  'List [ "config" ]': [ [Function: cursorRefresher], [Function: listener] ],
  'List [ "config", "profile" ]': [ [Function: cursorRefresher], [Function: listener] ] }
```

Is there a reason for this? It's probably better to use `Immutable.Map` as a tree to optimize the constant factor in the lookup operation; and use sentinel values (e.g. `LISTENERS = {}`) to retrieve the listener set.

Looking at https://github.com/omniscientjs/immstruct/pull/60/files#diff-dc0aa2a7df4698297c4ccd9dc0b071caR74, merging `Immutable.Set` of listeners at every subset of `keyPath` doesn't seem like it would scale for long `keyPaths` imo.

------

However, when investigating https://github.com/omniscientjs/immstruct/issues/37 I've thought up a similar solution.

But, I've come to the conclusion that it's best to actually separate the listeners into their respective contexts.
To elaborate, at `keyPath`, there would be two types of listeners:

1. listeners observing `keyPath`, and
2. listeners observing at path where `keyPath` is a subpath; that is listeners subscribed to the sub-tree of `keyPath`.

Then, on `swap`, you traverse `keyPath` to notify listeners in (1) in `this._referencelisteners`.
When you finally reached the end, just notify listeners in (2) at `keyPath`.

There are ways to optimize the second step to avoid traversing all the listeners in (2).

**Comment by tomasd** [comment]
I've pushed optimized version. It gives about 160 fps on my machine (original master was 170fps).

Listeners are now subscribed into path tree index and called when their path has changed.

**Comment by andrewluetgers** [comment]
Also BRAVO on the refactor :-)

**Comment by tomasd** [comment]
I've updated the code according to last comments and squashed the commits. Do you need something more for merge?

**Comment by mikaelbr** [comment]
Great work! I'd still prefer to have the dist to the current released version, not the unstable version. Could you undo the dist?

**Comment by tomasd** [comment]
Yes, I've just reverted the dist. Hope it's ok now :-)

**Comment by mikaelbr** [comment]
LGTM. There are some code stylistic changes I'd like to do, but I can refactor the code.
@andrewluetgers what do you think? Did you performance test the new changes?

@Dashed Are you happy with the PR? So #37 and #57 will be fixed for this. Are there any tests that we need to add?

**Comment by andrewluetgers** [comment]
@mikaelbr looks good there is no difference in performance now 
http://plnkr.co/edit/FMOIeJ7edrDaTimJ9tst?p=preview

**Comment by mikaelbr** [comment]
Perfect. I did some changes, refactoring and fixed some issues and added more tests. Merged now! Thanks for all your great work, @tomasd , @andrewluetgers  and @Dashed! :cake: 

This is a great feature!

File: src/structure.js [hunk 7]
```diff
@@ -66,21 +66,65 @@ function Structure (options) {
...
 }
 inherits(Structure, EventEmitter);
 module.exports = Structure;
+var FUNCTION_KEY = {};
+
+function emit(map, newData, oldData, path, args) {
+
+    if (map && newData !== oldData) {
+        map.get(FUNCTION_KEY, []).forEach(function (fn) {
+            fn.apply(null, args);
```
Comment (Dashed) [inline src/structure.js]: Shouldn't the `args` change with respect to `keyPath`?

Comment (tomasd) [inline src/structure.js]: It depends...In original version listeners were called with arguments from swap so I kept it.

Comment (Dashed) [inline src/structure.js]: Previously, there was no event propagation.

If I do something like this:
```js
var ref = structure.reference(['someBox']);

var unobserve = ref.observe(function () { 
});
```

It means I want to observe specifically at "someBox" path. This may lead to issues if the observing function is pure, and expect only changes at "someBox" path.

Comment (mikaelbr) [inline src/structure.js]: We should probably pass the "calculated" `keyPath`?

> This may lead to issues if the observing function is pure, and expect only changes at "someBox" path.

But it will only trigger when `someBox` or any of `someBox`s children has changed? And if children of `someBox` has changed, `someBox` has changed, no?

Comment (Dashed) [inline src/structure.js]: My interpretation was that it would trigger only when `someBox` path has truly changed; regardless of whether the change occurred at `someBox` path or at the parent of `someBox`.

Ideally, this would be how changes would be communicated:
```js
var ref = structure.reference(['someBox']);

var unobserve = ref.observe(function (newValue, oldValue) { 
});
```
